clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8927,8 +8927,13 @@ void Sema::CheckVariableDeclarationType(VarDecl *NewVD) {
     }
   }
 
-  if (T->isRVVSizelessBuiltinType())
-    checkRVVTypeSupport(T, NewVD->getLocation(), cast<Decl>(CurContext));
+  if (T->isRVVSizelessBuiltinType() && isa<FunctionDecl>(CurContext)) {
+    const FunctionDecl *FD = cast<FunctionDecl>(CurContext);
+    llvm::StringMap<bool> CallerFeatureMap;
+    Context.getFunctionFeatureMap(CallerFeatureMap, FD);
```

Comment:
Does this impact compile time if the RVV types are used many times in a function as they would be for intrinsic code?

Reply:
Yes, it does.
With 100k RVV type declarations,
This pr's compile time is approximately 8 times slower than original one.(0.56 seconds vs 0.07 seconds on the server)

Reply:
Mostly costed by `llvm::RISCVISAInfo::parseFeatures` in `RISCVTargetInfo::initFeatureMap`.

Reply:
Would it make sense to store a `Map<FunctionDecl *, StringMap<bool>` so `RISCVTargetInfo::initFeatureMap` only builds the feature map one time per FunctionDecl?

Reply:
But we don't have the `FunctionDecl` info in the `RISCVTargetInfo::initFeatureMap` call.

Reply:
However, if we can do something like that, we can get about 5x speedup with my rough measurement lol~

Reply:
@topperc I'm not sure why we also need "-" for features that are not enabled, is there any use case?

Reply:
@4vtomat if the MCPU has a feature but the function explicitly disables it then I think we want the `-`

Reply:
> But we don't have the FunctionDecl info in the RISCVTargetInfo::initFeatureMap call.

What about keeping this map in ASTContext instead?

Reply:
> @4vtomat if the MCPU has a feature but the function explicitly disables it then I think we want the `-`

Why don't we just remove it from feature map?

Reply:
> > But we don't have the FunctionDecl info in the RISCVTargetInfo::initFeatureMap call.
> 
> What about keeping this map in ASTContext instead?

I guess it might be a feasible solution, however the memory usage might increase since it has to keep the map for each function during the lifetime of the ASTContext.
@topperc Do you think we can do this? Or we can leave it in TODO until it becomes a problem in some real cases.

Reply:
> > @4vtomat if the MCPU has a feature but the function explicitly disables it then I think we want the `-`
> 
> Why don't we just remove it from feature map?

We also need the negative extensions if the target attribute string is a complete arch, e.g. `__attribute__((target="arch=rv64i")))`, because any extensions from mcpu need to be turned off.

Reply:
@lukel97 Got it, thanks!